### DISCUSSION TITLE
lua help - add pushd command

### DISCUSSION THREAD

**Discussion by eggbean** [chrisant996/clink#480] [Q&A] [answered]
Hi. I don't know lua and I always have trouble with it when I need to do something in a nvim plugin (I am totally fine with vimscript).

I am using this [`zoxide.lua`](https://github.com/shunsambongi/clink-zoxide/blob/fix-cd-command/zoxide.lua) script to add zoxide support to clink/cmd, but I also use [`CD Deluxe`](https://github.com/m6z/cd-deluxe), which I really like and find very convenient.

For `zoxide.lua` to work with `CD Deluxe` I need to add a `pushd %cd%` command before `zoxide` changes directory. How can I insert that here?

```
-- =============================================================================
--
-- Define aliases.
--

-- 'z' alias
local function __zoxide_z(keywords)
  if #keywords == 0 then
    return __zoxide_cd(os.getenv 'USERPROFILE')
  elseif #keywords == 1 then
    local keyword = keywords[1]
    if keyword == '-' then
      return __zoxide_cd '-'
    elseif os.isdir(keyword) then
      return __zoxide_cd(keyword)
    end
  end

  local cwd = '"' .. os.getcwd() .. '"'
  return __zoxide_query({ '--exclude', cwd }, keywords)
end

-- 'zi' alias
local function __zoxide_zi(keywords)
  return __zoxide_query({ '--interactive' }, keywords)
end

-- =============================================================================
```

I think there needs to be a `pushd %cd%` before each of the two `return __zoxide_cd` commands as well as the `return __zoxide_query` command.

I thought that adding this line before those command would work, but it breaks the whole thing:

    return pushd(os.getenv 'CD')

**Comment by chrisant996** [comment, marked as answer]
Oh, it looks like you're trying to use Lua as though it's a shell scripting language. It is not. `pushd` is a shell command, not a Lua function.

And `__zoxide_query` and `__zoxide_cd` are functions defined in the Lua script, and they return a string.  The string is returned to Clink, and Clink returns it to CMD, which executes the string as though the user typed it into the command prompt.

And `__zoxide_query` calls `__zoxide_cd`.

So you probably just need to change the `__zoxide_cd` function to return whatever command line you want to be executed.  It sounds like you want to put two commands in the command line, something like `" pushd %cd% & " ..` preceding whatever normally gets returned.

**Comment by eggbean** [comment]
Ah, I see. When compounding cmd command with `%` how do I stop doskey macros from expanding? [Now I am back to this problem](https://github.com/shunsambongi/clink-zoxide/issues/5#issuecomment-1493220327). Thanks.

```
local function __zoxide_cd(dir)
  if os.getenv '_ZO_ECHO' == '1' then
    print(dir)
  end

  -- 'cd /d -' doesn't work for clink versions before v1.2.41 (https://github.com/chrisant996/clink/issues/191)
  -- lastest cmder release (v1.3.18) uses clink v1.1.45
  if dir == '-' and (clink.version_encoded or 0) < 10020042 then
    return ' pushd %cd% & cd -'
  end

  return ' pushd %cd% & cd /d ' .. dir
end
```

**Reply by chrisant996** [reply]
Add a space.

Since it's typical to put a space after `&`, a second space is needed to avoid doskey expansion.

(It's mentioned in passing [here](https://chrisant996.github.io/clink/clink.html#line_state:getcommandoffset). I thought the enhanced doskey expansion rules were documented somewhere, but maybe not; I'll add that.)

Or another way is to [use a semicolon](https://chrisant996.github.io/clink/clink.html#saved-command-history) to disable doskey expansion.

**Comment by eggbean** [comment]
Oh, I just needed to add another space after the `&`. Thanks for the help.
